### PR TITLE
Add which entry command

### PR DESCRIPTION
**Pull request by ooye-sanket**
Adds `brew which-entry <formula>` a developer command that generates
a single `executables.txt` entry for a formula by reading
`sh.brew.path_exec_files` from its bottle manifest.

This is intended to replace the batch `brew which-update` job with a
per-formula update triggered by homebrew/core CI on each merge.
**What it does**
- Fetches the bottle manifest and reads `sh.brew.path_exec_files`
- With `--output-db=<file>`: upserts the formula's line in the DB file
  and removes it if the formula is disabled, deprecated, or deleted
- Without `--output-db=`: prints the DB line to stdout

**Relationship to other PRs**
The companion homebrew/core PR (Homebrew/homebrew-core#272976) slims
the existing workflow down to just call `brew which-entry` per formula
detected by `brew test-bot --only-formulae-detect`, replacing the
current daily batch job entirely.

@MikeMcQuaid please have a look!

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew lgtm` (style, typechecking and tests) with your changes locally?

-----

- [ ] AI was used to generate or assist with generating this PR. *Please specify below how you used AI to help you, and what steps you have taken to manually verify the changes*.

-----

### CODE REVIEW

File: Library/Homebrew/dev-cmd/which-entry.rb [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        manifest_path = cached_manifest(formula) || fetch_manifest(formula)
+        return [] unless manifest_path
+
+        manifest = JSON.parse(manifest_path.read)
+        exec_files = manifest.dig("manifests", 0, "annotations", "sh.brew.path_exec_files")
+        return [] if exec_files.blank?
+
+        exec_files.split.map { |f| File.basename(f) }.sort
+      rescue JSON::ParserError
+        []
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/which-entry.rb]: Probably worth a `opoo` warning of the exception in this case?

**Review by MikeMcQuaid — COMMENTED** [review]
Nice work @ooye-sanket! Looks good so far. Please add a very basic integration test and args check for `which-entry`. Also, please inline functions that are only used once as there's a fair bit of indirection right now.

**Comment by ooye-sanket** [comment]
Thanks @MikeMcQuaid! I've pushed the changes:

- Inlined `cached_manifest`, `fetch_manifest`, `read_db`, and `remove_entry` .
- Added `opoo` warning when `JSON::ParserError` is raised so failures aren't silent
- Added the basic spec with args check

Let me know if the level of inlining looks right or if you'd like it structured differently!

File: Library/Homebrew/dev-cmd/which-entry.rb [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+          Generate an `executables.txt` entry for <formula> using bottle manifest metadata.
+        EOS
+        flag "--output-db=",
+             description: "Append or update the entry in the given `executables.txt` database file."
+        named_args :formula, min: 1
+      end
+
+      sig { override.void }
+      def run
+        db_path = args.output_db&.then { |path| Pathname(path) }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/which-entry.rb]: Not sure I understand the `.then` usage here: can you explain?

File: Library/Homebrew/dev-cmd/which-entry.rb [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+      end
+
+      private
+
+      sig { params(name: String, db_path: T.nilable(Pathname)).void }
+      def process(name, db_path:)
+        formula = Formulary.factory(name)
+        line = db_line(formula)
+        if db_path
+          existing = db_path.exist? ? db_path.readlines(chomp: true).reject(&:empty?) : []
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/which-entry.rb]:
```suggestion
          existing = db_path.readlines(chomp: true).reject(&:blank?) if db_path.exist?
          existing ||= []
```
maybe

File: Library/Homebrew/dev-cmd/which-entry.rb [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+          lines = existing.reject { |l| l.start_with?("#{formula.full_name}(") }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/which-entry.rb]: Could maybe drop `existing` as a variable and just chain this?

File: Library/Homebrew/dev-cmd/which-entry.rb [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+          lines << line if line
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/which-entry.rb]: Maybe use `filter_map`?

File: Library/Homebrew/dev-cmd/which-entry.rb [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+          db_path.write("#{lines.sort.join("\n")}\n")
+        else
+          puts line if line
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/which-entry.rb]:
```suggestion
        elsif line
          puts line
```

File: Library/Homebrew/dev-cmd/which-entry.rb [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        end
+      rescue FormulaUnavailableError
+        return unless db_path&.exist?
+
+        lines = db_path.readlines(chomp: true).reject { |l| l.start_with?("#{name}(") }
+        db_path.write("#{lines.sort.join("\n")}\n")
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/which-entry.rb]: This logic is repeated, maybe worth turning into a function or restructuring

**Review by MikeMcQuaid — COMMENTED** [review]
Thanks! Please ensure CI syntax job passing before asking for re-review.

File: Library/Homebrew/dev-cmd/which-entry.rb [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+          Generate an `executables.txt` entry for <formula> using bottle manifest metadata.
+        EOS
+        flag "--output-db=",
+             description: "Append or update the entry in the given `executables.txt` database file."
+        named_args :formula, min: 1
+      end
+
+      sig { override.void }
+      def run
+        db_path = Pathname(T.unsafe(args).output_db) if T.unsafe(args).output_db
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/which-entry.rb:23]: This shouldn't be necessary, you may need to run `brew typecheck --update` to avoid the need for `T.unsafe`.

File: Library/Homebrew/dev-cmd/which-entry.rb [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+          puts line
+        end
+      rescue FormulaUnavailableError
+        write_db(db_path, name, nil) if db_path&.exist?
+      end
+
+      sig { params(db_path: Pathname, name: String, line: T.nilable(String)).void }
+      def write_db(db_path, name, line)
+        lines = db_path.readlines(chomp: true).compact_blank if db_path.exist?
+        lines = (lines || []).reject { |l| l.start_with?("#{name}(") }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/which-entry.rb:45]:
```suggestion
        lines = Array(lines).reject { |l| l.start_with?("#{name}(") }
```

File: Library/Homebrew/dev-cmd/which-entry.rb [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+        manifest = JSON.parse(manifest_path.read)
+        exec_files = manifest.dig("manifests", 0, "annotations", "sh.brew.path_exec_files")
+        return [] if exec_files.blank?
+
+        exec_files.split.map { |f| File.basename(f) }.sort
+      rescue JSON::ParserError => e
+        opoo "Failed to parse bottle manifest for #{formula.name}: #{e.message}"
+        []
+      end
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/which-entry.rb:80]: can this share any logic or constants with the relevant `bottle.rb` or `resource.rb` code?

**Review by MikeMcQuaid — COMMENTED** [review]
Thanks, looking good, a few small things!

File: Library/Homebrew/dev-cmd/which-entry.rb [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+        lines << line if line
+        db_path.write("#{lines.sort.join("\n")}\n")
+      end
+
+      sig { params(formula: Formula).returns(T.nilable(String)) }
+      def db_line(formula)
+        return if formula.disabled? || formula.deprecated?
+
+        exes = executables_from_manifest(formula)
+        "#{formula.full_name}(#{formula.pkg_version}):#{exes.join(" ")}"
```
Comment (Rylan12) [inline Library/Homebrew/dev-cmd/which-entry.rb:55]: The reason we added `pkg_version` before was to track formula file changes over time (we only checked the formulae with changed versions for updated exes).

Now that we're going to run this on every merge where we can detect which formulae are updated, I don't think we need the pkg_version anymore

**Review by Rylan12 — COMMENTED** [review]
Sorry for the review delay here. I agree with Mike's feedback, and also have one other thought.

More generally, I'm not sure whether this actually needs to be a `brew` command at all. I will leave a comment on the other PR with more details

Edit: oh, one more thing, we should remove the existing [`command-not-found-db-update.yml`](https://github.com/Homebrew/brew/blob/main/.github/workflows/command-not-found-db-update.yml) workflow

**Comment by ooye-sanket** [comment]
hey @Rylan12  The reason for keeping it as a brew command is that it already handles formula loading, bottle manifest caching, disabled/deprecated checks, and DB file management in a tested, typed Ruby context. Doing all that reliably in bash would be fragile. But happy to discuss if there's a simpler scope you have in mind.

**Comment by MikeMcQuaid** [comment]
Let's keep it as Ruby for now and we can see how it's looking when the other PR is ready.

**Comment by Rylan12** [comment]
I don't mind keeping it in Ruby for now, so this is not a blocker.

> it already handles formula loading, bottle manifest caching, disabled/deprecated checks, and DB file management in a tested, typed Ruby context. Doing all that reliably in bash would be fragile

I think doing it in bash would actually let us bypass these complexities altogether, resulting in a simpler system overall. There would be no need to load formulae at all, and the deprecated/disabled info is available via the same bottle manifest that we already need to use.

Fair enough on the DB management, although I think a reasonably well-crafted `sed` line would do the trick.

I haven't attempted to write it out, so I could be wrong

**Comment by ooye-sanket** [comment]
Hey @MikeMcQuaid I've made all the changes (dropped pkg_version, fixed the reject pattern, removed T.unsafe) but I'm unable to run brew typecheck --update locally to regenerate the RBI file. Could you help run it or point me to how to get past this? Everything else should be good to go!

**Comment by MikeMcQuaid** [comment]
@ooye-sanket what happens when you run it locally?

**Comment by MikeMcQuaid** [comment]
@ooye-sanket nudge on above question!

**Review by MikeMcQuaid — COMMENTED** [review]
Waiting on author answering question.

**Comment by ooye-sanket** [comment]
Hey @MikeMcQuaid, sorry for the delay! I ran brew typecheck --update locally and it worked successfully it generated the new sorbet/rbi/dsl/homebrew/dev_cmd/which_entry.rbi RBI file which I've now committed. I've also rebased onto latest upstream/main. Ready for re-review

Note: the brew typecheck CI failure is in extend/os/mac/keg_relocate.rb and extend/os/mac/keg.rb these are pre-existing errors unrelated to this PR.

CC @Rylan12

**Comment by MikeMcQuaid** [comment]
> Note: the brew typecheck CI failure is in extend/os/mac/keg_relocate.rb and extend/os/mac/keg.rb these are pre-existing errors unrelated to this PR.

That's not true: these are passing on `main`: https://github.com/Homebrew/brew/actions/runs/24983715091/job/73151721181

**Review by MikeMcQuaid — COMMENTED** [review]
Still waiting on author

**Comment by ooye-sanket** [comment]
Hey @MikeMcQuaid, CI is now green! , The earlier typecheck failure was caused by our update accidentally removing mach_o_shim.rbi  I've now restored it. Ready for re-review! 🙏

File: Library/Homebrew/dev-cmd/which-entry.rb (hunk 8 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/dev-cmd/which-entry.rb]: @ooye-sanket Can you take another look at this?

File: Library/Homebrew/dev-cmd/which-entry.rb (hunk 11 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/dev-cmd/which-entry.rb]: @ooye-sanket and this

File: Library/Homebrew/dev-cmd/which-entry.rb (hunk 10 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/dev-cmd/which-entry.rb:66]: @ooye-sanket and this

File: Library/Homebrew/sorbet/rbi/dsl/cask/config.rbi [hunk 12]
```diff
@@ -10,45 +10,12 @@ module Cask
     sig { returns(String) }
     def appdir; end
 
-    sig { returns(String) }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/sorbet/rbi/dsl/cask/config.rbi:13]: These changes seem unrelated to the PR?

**Comment by ooye-sanket** [comment]
Hey @MikeMcQuaid, sorry for the noise! All 3 issues fixed.
CI is green and ready for re-review!

File: Library/Homebrew/dev-cmd/which-entry.rb [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+          Generate an `executables.txt` entry for <formula> using bottle manifest metadata.
+        EOS
+        flag "--output-db=",
+             description: "Append or update the entry in the given `executables.txt` database file."
+        named_args :formula, min: 1
+      end
+
+      sig { override.void }
+      def run
+        db_path = Pathname(T.must(args.output_db)) if args.output_db
```
Comment (Rylan12) [inline Library/Homebrew/dev-cmd/which-entry.rb]: IMO we should just make this argument required and raise a usage error if it's not passed. Or accept it as a named arg

File: Library/Homebrew/dev-cmd/which-entry.rb (hunk 9 above)
Comment (Rylan12) [reply Library/Homebrew/dev-cmd/which-entry.rb]: This is not resolved

Comment (Rylan12) [reply Library/Homebrew/dev-cmd/which-entry.rb]: Also, there's an inconsistency here, right? We reject lines that start with `name(` but we don't actually write the version in parentheses anymore.

File: Library/Homebrew/dev-cmd/which-entry.rb [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+
+        manifest_path = HOMEBREW_CACHE.glob("#{formula.name}_bottle_manifest--*").first
+        if manifest_path.blank?
+          bottle = formula.bottle
+          return [] unless bottle
+
+          bottle.fetch
+          manifest_path = HOMEBREW_CACHE.glob("#{formula.name}_bottle_manifest--*").first
+        end
+        return [] if manifest_path.blank?
```
Comment (Rylan12) [inline Library/Homebrew/dev-cmd/which-entry.rb]: We should be able to access this using `formula.bottle.github_packages_manifest_resource`. This is a `Resource` object that includes `Downloadable`, and there should be lots of methods in that class that will help avoid the filesystem glob like `downloaded?` and `cached_location`. We should use these methods instead of globbing

**Comment by Rylan12** [comment]
@ooye-sanket, some of the issues Mike pointed out don't actually seem to have been fixed yet. When you fix those, can you please run the commands locally on your machine and comment the output here to help verify that this works as expected?

It would be helpful to see a sample `executables.txt` with pkg_versions and and outdated exe list, then show the `brew which-entry` command and it's output, and then the fixed `executables.txt` file

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]
Comments not addressed. Do not mark as resolved or request re review until 1) all comments past and present resolved and 2) all @Rylan12's comments and questions answered.

If I'm requested for review again and this still isn't ready: I'm just likely to close this out, sorry. Please raise the bar for yourself here. Thank you ❤️

File: Library/Homebrew/dev-cmd/which-entry.rb (hunk 13 above)
Comment (ooye-sanket) [reply Library/Homebrew/dev-cmd/which-entry.rb]: Fixed --output-db is now required and raises a UsageError if not passed.

File: Library/Homebrew/dev-cmd/which-entry.rb (hunk 9 above)
Comment (ooye-sanket) [reply Library/Homebrew/dev-cmd/which-entry.rb]: Fixed as now uses Array(lines) and the reject pattern is corrected to "#{name}:" to match the new format without pkg_version.

File: Library/Homebrew/dev-cmd/which-entry.rb (hunk 14 above)
Comment (ooye-sanket) [reply Library/Homebrew/dev-cmd/which-entry.rb]: Fixed now uses formula.bottle.github_packages_manifest_resource with downloaded? and cached_download instead of filesystem globbing.

**Comment by ooye-sanket** [comment]
Hey @MikeMcQuaid and @Rylan12, all comments are now addressed in the latest commits.
- --output-db is now required raises UsageError if not passed
- write_db now uses Array(lines) and correctly rejects name: instead of name(
- executables_from_manifest now uses formula.bottle.github_packages_manifest_resource with downloaded? and   cached_download no more filesystem globbing
- Removed command-not-found-db-update.yml as requested
### **Here's proof the command works locally:**
{136E2976-0335-4DD9-8772-C398249919C7}
{88692EFA-6CF5-495F-9132-113ADF4D0787}
Entries are sorted, idempotent, and correctly formatted as name:executables.

NOTE : The only failing CI check is Docker / docker (x86_64 Ubuntu 22.04) which is a GitHub infrastructure issue apt-get update failing with permission errors, unrelated to our code. All 32 other checks are passing

File: Library/Homebrew/dev-cmd/which-entry.rb [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+
+      private
+
+      sig { params(name: String, db_path: Pathname).void }
+      def process(name, db_path:)
+        formula = Formulary.factory(name)
+        line = db_line(formula)
+        write_db(db_path, formula.full_name, line)
+      rescue FormulaUnavailableError
+        write_db(db_path, name, nil) if db_path.exist?
```
Comment (Rylan12) [inline Library/Homebrew/dev-cmd/which-entry.rb]: Why are we rescuing here? Seems like we should just raise the FormulaUnavailableError?

Or better yet, use `args.named.to_formulae` or something like that?

File: Library/Homebrew/dev-cmd/which-entry.rb [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+      end
+
+      sig { params(db_path: Pathname, name: String, line: T.nilable(String)).void }
+      def write_db(db_path, name, line)
+        lines = db_path.readlines(chomp: true).compact_blank if db_path.exist?
+        lines = Array(lines).reject { |l| l.start_with?("#{name}:") }
```
Comment (Rylan12) [inline Library/Homebrew/dev-cmd/which-entry.rb]:
```suggestion
        # TODO: remove once pkg_versions are no longer in executables.txt
        lines = Array(lines).reject { |l| l.start_with?(/#{name}(?:\(.+\))?:/) }
```

Not tested, so please verify that the regex works as expected

File: Library/Homebrew/dev-cmd/which-entry.rb [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+
+        bottle = formula.bottle
+        return [] unless bottle
+
+        manifest_resource = bottle.github_packages_manifest_resource
+        return [] unless manifest_resource
+
+        manifest_resource.fetch unless manifest_resource.downloaded?
+        manifest_path = manifest_resource.cached_download
+        return [] if manifest_path.blank?
```
Comment (Rylan12) [inline Library/Homebrew/dev-cmd/which-entry.rb]:
```suggestion
        manifest_resource = formula.bottle&.github_packages_manifest_resource
        return [] unless manifest_resource

        manifest_resource.fetch unless manifest_resource.downloaded?
        manifest_path = manifest_resource.cached_download
        return [] unless manifest_path.exist?
```

Can consolidate this with chaining. Also, the `blank?` check doesn't work here because `blank?` returns `false` when files aren't present, so you'll hit an error on the next line

File: Library/Homebrew/dev-cmd/which-entry.rb [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+
+        manifest = JSON.parse(manifest_path.read)
+        exec_files = manifest.dig("manifests", 0, "annotations", "sh.brew.path_exec_files")
```
Comment (Rylan12) [inline Library/Homebrew/dev-cmd/which-entry.rb]: I think we need to be more careful with which manifest we ready from, since I don't think these are guaranteed to be in order. In my opinion, we should read from whichever annotation matches the current system, which will always be consistent on the GitHub actions runners.

Long term, we should maybe consider providing a different `executables.txt` file for each platform in case it's different (or maybe just a separate macOS/Linux version)

File: Library/Homebrew/dev-cmd/which-entry.rb (hunk 9 above)
Comment (Rylan12) [reply Library/Homebrew/dev-cmd/which-entry.rb]: You need to support both the old and new formats

File: Library/Homebrew/dev-cmd/which-entry.rb [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+             description: "Append or update the entry in the given `executables.txt` database file."
+        named_args :formula, min: 1
+      end
+
+      sig { override.void }
+      def run
+        raise UsageError, "`--output-db` is required." unless args.output_db
+
+        db_path = Pathname(T.must(args.output_db))
+        args.named.each { |name| process(name, db_path:) }
```
Comment (Rylan12) [inline Library/Homebrew/dev-cmd/which-entry.rb]:
```suggestion
        db_path = args.output_db
        raise UsageError, "`--output-db` is required." unless db_path

        args.named.each { |name| process(name, db_path: Pathname(db_path)) }
```

Something like this should be used to avoid `T.must`

File: Library/Homebrew/dev-cmd/which-entry.rb (hunk 15 above)
Comment (ooye-sanket) [reply Library/Homebrew/dev-cmd/which-entry.rb]: Fixed removed the rescue and now using args.named.to_formulae directly.

File: Library/Homebrew/dev-cmd/which-entry.rb (hunk 16 above)
Comment (ooye-sanket) [reply Library/Homebrew/dev-cmd/which-entry.rb]: Applied your suggestion using Regexp.escape(name) for safety too.

File: Library/Homebrew/dev-cmd/which-entry.rb (hunk 17 above)
Comment (ooye-sanket) [reply Library/Homebrew/dev-cmd/which-entry.rb]: Applied your suggestion consolidated with &. chaining and fixed blank? → exist?.

File: Library/Homebrew/dev-cmd/which-entry.rb (hunk 18 above)
Comment (ooye-sanket) [reply Library/Homebrew/dev-cmd/which-entry.rb]: Fixed extracted path_exec_files as a public method on Resource::BottleManifest which uses the existing manifest_annotations method that already correctly matches the current platform via bottle checksum and image ref.

File: Library/Homebrew/dev-cmd/which-entry.rb (hunk 19 above)
Comment (ooye-sanket) [reply Library/Homebrew/dev-cmd/which-entry.rb]: Applied your suggestion exactly.

**Comment by ooye-sanket** [comment]
Hey @Rylan12, all feedback addressed! please have a look 
image

File: Library/Homebrew/dev-cmd/which-entry.rb [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+      end
+
+      sig { params(formula: Formula).returns(T::Array[String]) }
+      def executables_from_manifest(formula)
+        manifest_resource = formula.bottle&.github_packages_manifest_resource
+        return [] unless manifest_resource
+
+        manifest_resource.fetch unless manifest_resource.downloaded?
+        manifest_path = manifest_resource.cached_download
+        return [] unless manifest_path.exist?
```
Comment (Rylan12) [inline Library/Homebrew/dev-cmd/which-entry.rb:60]: I _think_ `.fetch` is probably good enough here. It's called by default with `verify_download_integrity: true`, so I think we'll get an error before this point if we weren't able to download the manifest. I'd say we can just get rid of these lines since I'm not convinced they add additional safety. It seems appropriate to me for this to fail if the formula is bottled but unable to download the bottle manifest. That does seem error-worthy


```suggestion
```

File: Library/Homebrew/dev-cmd/which-entry.rb [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+
+        manifest_resource.path_exec_files
+      rescue JSON::ParserError => e
+        opoo "Failed to parse bottle manifest for #{formula.name}: #{e.message}"
+        []
```
Comment (Rylan12) [inline Library/Homebrew/dev-cmd/which-entry.rb:65]: Let's get rid of this. `BottleResource#manifest_annotations` raises a nice error if the JSON doesn't parse, and I think we want a hard error here. Also it's a bit weird to rescue a JSON parse error when we don't actually call `JSON.parse` directly in this method.


```suggestion
```

File: Library/Homebrew/resource.rb [hunk 22]
```diff
@@ -411,6 +411,14 @@ def installed_size
       manifest_annotations["sh.brew.bottle.installed_size"]&.to_i
     end
 
+    sig { returns(T::Array[String]) }
+    def path_exec_files
+      exec_files = manifest_annotations["sh.brew.path_exec_files"]
+      return [] if exec_files.blank?
+
+      exec_files.split.map { |f| File.basename(f) }.sort
+    end
```
Comment (Rylan12) [inline Library/Homebrew/resource.rb:420]: This does not work correctly for annotations that have multiple entries. The entries are comma-separated, so `split` doesn't work.

Additionally, the `File.basename` doesn't belong in this method. This method should return the data directly, and leave it up to the caller how to process it.


```suggestion
    def path_exec_files
      manifest_annotations.fetch("sh.brew.path_exec_files", "").split(",")
    end
```

**Review by Rylan12 — CHANGES_REQUESTED** [review]
@ooye-sanket, this implementation does not work correctly for any formulae that have multiple executables.

For future pushes, please include proof that you've tested the changes and that they work correctly for the standard case **and also edge cases**. Since this will run on every formula update, it's imperative that the code is sound and can handle all of the possible situations that may arise.

To help with this, unit tests would be useful for all changes (aside from a full integration tests, which are okay to skip because they're too slow).

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]
@ooye-sanket please stop requesting my review here, thanks. I'll review when @Rylan12 is ✅.
